File: kotlin-semver/karakum/plugins/convertComparisonResultType.js
import ts from "typescript";
import * as karakum from "karakum";

const comparisonResultType = JSON.stringify(["1", "0", "-1"].sort())

export default function (node, context) {
    const typeScriptService = context.lookupService(karakum.typeScriptServiceKey)

    if (
        ts.isUnionTypeNode(node)
        && node.types.every(type => ts.isLiteralTypeNode(type) && (ts.isNumericLiteral(type.literal) || (ts.isPrefixUnaryExpression(type.literal) && ts.isNumericLiteral(type.literal.operand))))
        && JSON.stringify(node.types.map(type => typeScriptService.printNode(type)).sort()) === comparisonResultType

        && node.parent
        && (ts.isFunctionDeclaration(node.parent) || ts.isMethodDeclaration(node.parent))
        && node.parent.type === node

        && ts.isIdentifier(node.parent.name)
        && (node.parent.name.text.startsWith("compare")
            || node.parent.name.text.startsWith("rcompare"))
    ) {
        return "Int"
    }

    return null
}

File: kotlin-semver/karakum/src/jsMain/kotlin/semver/karakum/plugins/convertComparisonResultType.kt
package semver.karakum.plugins

import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import io.github.sgrishchenko.karakum.extension.plugins.TypeScriptService
import io.github.sgrishchenko.karakum.extension.plugins.typeScriptServiceKey
import io.github.sgrishchenko.karakum.util.getParentOrNull
import typescript.*

private val comparisonResultType = listOf("1", "0", "-1")

val convertComparisonResultType = createPlugin { node, context, _ ->
    nullable {
        val typeScriptService = ensureNotNull(context.lookupService(typeScriptServiceKey))

        ensure(isUnionTypeNode(node))
        ensure(node.types.asArray().all { type ->
            nullable {
                ensure(isLiteralTypeNode(type))

                val literal = type.literal

                nullable {
                    ensure(isNumericLiteral(literal))
                } ?: nullable {
                    ensure(isPrefixUnaryExpression(literal))

                    val operand = literal.operand
                    ensure(isNumericLiteral(operand))
                }
            } != null
        })

        val printedTypes = node.types.asArray().map { typeScriptService.printNode(it) }
        ensure(printedTypes == comparisonResultType)

        val parent = ensureNotNull(node.getParentOrNull())

        val name = ensureNotNull(
            nullable {
                ensure(isFunctionDeclaration(parent))

                parent.name
            } ?: nullable {
                ensure(isMethodDeclaration(parent))

                val methodName = parent.name
                ensure(isIdentifier(methodName))

                methodName
            }
        )
        ensure(
            name.text.startsWith("compare")
                    || name.text.startsWith("rcompare")
        )

        "Int"
    }
}
